pyquil/operator_estimation.py

```diff
@@ -479,3 +497,164 @@ def measure_observables(qc: QuantumComputer, tomo_experiment: TomographyExperime
+            appropriately into it
+    """
+    # construct a tuple out of the in and out operators of the ExperimentSetting
+    tup_pauli_terms = tuple(([expt_setting.in_operator], [expt_setting.out_operator]))
+    # loop through keys and determine if in_operator is diagonal in the same tpb as in_key,
+    # and out_operator the same with out_key
+    for key in list(diagonal_sets.keys()):
+        # obtain the in and out keys
+        in_key = key[0]
+        out_key = key[1]
+        # construct PauliTerms from in and out keys
+        pauli_in_from_key = PauliTerm.from_list([x[::-1] for x in in_key])
+        pauli_out_from_key = PauliTerm.from_list([x[::-1] for x in out_key])
+        # determine if in_operator is diagonal in the same tpb as the PauliTerm constructed
+        # from the in_key; do the same for the out_operator and the out_key
+        b_in_commutes = diagonal_basis_commutes(tup_pauli_terms[0][0], pauli_in_from_key)
+        b_out_commutes = diagonal_basis_commutes(tup_pauli_terms[1][0], pauli_out_from_key)
+        # update the dict value if both the pairs are diagonal in the same tpb
+        # (note: the two tpbs need not be the same)
+        if b_in_commutes and b_out_commutes:
+            # updated the in and out sets comprising the dict's value
+            updated_pauli_in_set = diagonal_sets[key][0] + tup_pauli_terms[0]
+            updated_pauli_out_set = diagonal_sets[key][1] + tup_pauli_terms[1]
+            updated_pauli_set = (updated_pauli_in_set, updated_pauli_out_set)
+            # obtain the diagonalizing bases for both the updated in and out sets
+            diagonalizing_in_term = get_diagonalizing_basis(updated_pauli_in_set)
+            diagonalizing_out_term = get_diagonalizing_basis(updated_pauli_out_set)
+            # update the diagonalizing basis (key of dict) if necessary
+            if len(diagonalizing_in_term) > len(in_key) or len(diagonalizing_out_term) > len(out_key):
+                del diagonal_sets[key]
+                new_in_key = tuple(sorted(diagonalizing_in_term._ops.items(), key=lambda x: x[0]))
+                new_out_key = tuple(sorted(diagonalizing_out_term._ops.items(), key=lambda x: x[0]))
+                new_key = (new_in_key, new_out_key)
+                diagonal_sets[new_key] = updated_pauli_set
+            else:
+                diagonal_sets[key] = updated_pauli_set
+            return diagonal_sets
+
+    # made it through all keys and sets so need to make a new set
+    # always need to sort because new pauli term functionality
+    new_in_key = tuple(sorted(tup_pauli_terms[0][0]._ops.items(), key=lambda x: x[0]))
+    new_out_key = tuple(sorted(tup_pauli_terms[1][0]._ops.items(), key=lambda x: x[0]))
+    new_key = (new_in_key, new_out_key)
+    diagonal_sets[new_key] = tup_pauli_terms
+    return diagonal_sets
+
+
+def commuting_sets_by_zbasis_tomo_expt(tomoexpt: TomographyExperiment):
```

Comment:
this function should be private, ie prepend an underscore to its name

Reply:
Agreed. Updated.